In [26]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder,StandardScaler,LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split,GridSearchCV,cross_val_predict
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix,mean_squared_error,r2_score,make_scorer
from sklearn.model_selection import cross_val_score,KFold
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.decomposition import PCA

In [27]:
# Load the house price data
data = pd.read_csv('C:\\Users\\91934\\OneDrive\\Desktop\\GMU\\Ait 664\\Hands on session\\Boston-house-price-data.csv')
print(data.info())
print(data.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB
None
             CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695   

In [28]:
missing=data.isnull().sum()
missing

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

In [29]:

#Create the target variable based on the median price
#Class 1: house prices >= median, Class 0: house prices < median
median_price =data['MEDV'].median()
data['price_category'] = (data['MEDV'] >= median_price).astype(int)
# Separate the target variable
X = data.drop(['price_category','MEDV'],axis=1)
y = data['price_category']
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
data

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  price_category  
0       15.3  396.90   4.98  24.0               1  
1       17.8  396.90   9.14  21.6               1  
2       17.8  392.83   4.03  34.7               1  
3       18.7  394.63   2.94  33.4               1  
4       18.7  396.90   5.33  36.2               1  
..       ...     ...    ...   ...             ...  
501     21.0  391.99   9.67  22.4               1  
502     21.0  396.90   9.08  20.6               0  
503     21.0  396.90   5.64  23.9               1  
504     21.0  393.45   6.48  22.0               1  
505     21.0  396.90   7.88  11.9               0  

[506 rows x 15 columns]

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [31]:
# Initialize the Logistic Regression model
logistic_model = LogisticRegression(max_iter=100, random_state=42)
# Fit the model on the training data
logistic_model.fit(X_train, y_train)
# Predict on the testing set
y_pred = logistic_model.predict(X_test)

c:\Users\91934\AppData\Local\Programs\Python\Python38\lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [32]:
# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")
# Display a classification report for precision, recall, and F1-score
print("Classification Report:\n", classification_report(y_test, y_pred))
# Display a confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))


Accuracy: 0.83
Classification Report:
               precision    recall  f1-score   support

           0       0.91      0.79      0.85        90
           1       0.74      0.89      0.81        62

    accuracy                           0.83       152
   macro avg       0.83      0.84      0.83       152
weighted avg       0.84      0.83      0.83       152

Confusion Matrix:
 [[71 19]
 [ 7 55]]


In [33]:
# Initialize the DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(criterion='gini', max_depth=3, random_state=42)
# Fit the model on the training data
decision_tree.fit(X_train, y_train)
# Predict on the testing set
y_pred = decision_tree.predict(X_test)
predictions=pd.DataFrame({'Actual':y_test,'Predicted':y_pred})
predictions

Actual  Predicted
173       1          1
274       1          1
491       0          0
72        1          1
452       0          0
..      ...        ...
441       0          0
23        0          0
225       1          1
433       0          0
447       0          0

[152 rows x 2 columns]

In [34]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

# Display a classification report for precision, recall, and F1-score
print("Classification Report:\n", classification_report(y_test, y_pred))

# Display a confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.82
Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.79      0.84        90
           1       0.74      0.87      0.80        62

    accuracy                           0.82       152
   macro avg       0.82      0.83      0.82       152
weighted avg       0.83      0.82      0.82       152

Confusion Matrix:
 [[71 19]
 [ 8 54]]


In [35]:
#Apply random forest model
random_forest = RandomForestClassifier(n_estimators=1, criterion='gini', max_depth=None, random_state=42)
random_forest.fit(X_train, y_train)
#make predictions
y_pred = random_forest.predict(X_test)

In [36]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")
# Display a classification report for precision, recall, and F1-score
print("Classification Report:\n", classification_report(y_test, y_pred))
# Display a confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.76
Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.73      0.79        90
           1       0.68      0.81      0.74        62

    accuracy                           0.76       152
   macro avg       0.76      0.77      0.76       152
weighted avg       0.78      0.76      0.77       152

Confusion Matrix:
 [[66 24]
 [12 50]]


In [37]:
# Initialize the GradientBoostingClassifier
gbm = GradientBoostingClassifier(
    n_estimators=100,    # Number of boosting stages (trees)
    learning_rate=0.05,   # Step size for updating weights (lower values make learning slower)
    max_depth=5,         # Maximum depth of individual trees
    random_state=42
)

# Fit the model on the training data
gbm.fit(X_train, y_train)
# Predict on the testing set
y_pred = gbm.predict(X_test)

In [38]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

# Display a classification report for precision, recall, and F1-score
print("Classification Report:\n", classification_report(y_test, y_pred))

# Display a confusion matrix
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.85
Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.80      0.86        90
           1       0.76      0.92      0.83        62

    accuracy                           0.85       152
   macro avg       0.85      0.86      0.85       152
weighted avg       0.86      0.85      0.85       152

Confusion Matrix:
 [[72 18]
 [ 5 57]]


In [39]:
#set k values initially to 4
k = 4
#apply define model
knn_model = KNeighborsClassifier(n_neighbors=k)
#apply to the trainig data
knn_model.fit(X_train, y_train)
#make predictions using the test data
y_pred = knn_model.predict(X_test)

In [40]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

Accuracy: 0.81
Classification Report:
               precision    recall  f1-score   support

           0       0.88      0.79      0.83        90
           1       0.73      0.84      0.78        62

    accuracy                           0.81       152
   macro avg       0.80      0.81      0.81       152
weighted avg       0.82      0.81      0.81       152

Confusion Matrix:
 [[71 19]
 [10 52]]


Conclusion:
Logistic Regression- 0.83
Decision Tree-0.82
Random Forest-0.76
Gradient Boost-0.85
KNN- 0.81